src/vs/base/browser/ui/iconLabel/iconLabel.ts

```diff
@@ -155,18 +155,86 @@ export class IconLabel extends Disposable {
 
 			if (this.descriptionNode instanceof HighlightedLabel) {
 				this.descriptionNode.set(description || '', options ? options.descriptionMatches : undefined);
-				if (options?.descriptionTitle) {
-					this.descriptionNode.element.title = options.descriptionTitle;
-				} else {
-					this.descriptionNode.element.removeAttribute('title');
-				}
+				this.setupHover(this.descriptionNode.element, options?.descriptionTitle);
 			} else {
 				this.descriptionNode.textContent = description || '';
-				this.descriptionNode.title = options?.descriptionTitle || '';
+				this.setupHover(this.descriptionNode.element, options?.descriptionTitle || '');
 				this.descriptionNode.empty = !description;
 			}
 		}
 	}
+
+	private setupHover(htmlElement: HTMLElement, tooltip: string | IMarkdownString | Promise<IMarkdownString | string | undefined> | undefined): void {
+		if (this.customHovers.has(htmlElement)) {
+			this.customHovers.get(htmlElement)!.dispose();
+			this.customHovers.delete(htmlElement);
+		}
+
+		if (!tooltip) {
+			htmlElement.removeAttribute('title');
+			return;
+		}
+
+		if (!this.hoverDelegate) {
+			return this.setupNativeHover(htmlElement, tooltip);
+		} else {
+			return this.setupCustomHover(this.hoverDelegate, htmlElement, tooltip);
+		}
+	}
+
+	private setupCustomHover(hoverDelegate: IHoverDelegate, htmlElement: HTMLElement, tooltip: string | IMarkdownString | Promise<IMarkdownString | string | undefined> | undefined): void {
+		htmlElement.removeAttribute('title');
+		// Testing has indicated that on Windows and Linux 500 ms matches the native hovers most closely.
+		// On Mac, the delay is 1500.
+		const hoverDelay = isMacintosh ? 1500 : 500;
+		let hoverOptions: IHoverDelegateOptions | undefined;
+		let mouseX: number | undefined;
+		function mouseOver(this: HTMLElement, e: MouseEvent): any {
+			let isHovering = true;
+			function mouseMove(this: HTMLElement, e: MouseEvent): any {
+				mouseX = e.x;
+			}
+			function mouseLeave(this: HTMLElement, e: MouseEvent): any {
+				isHovering = false;
+			}
+			const mouseLeaveDisposable = domEvent(htmlElement, dom.EventType.MOUSE_LEAVE, true)(mouseLeave.bind(htmlElement));
+			const mouseMoveDisposable = domEvent(htmlElement, dom.EventType.MOUSE_MOVE, true)(mouseMove.bind(htmlElement));
+			setTimeout(async () => {
+				if (isHovering && tooltip) {
+					// Re-use the already computed hover options if they exist.
+					if (!hoverOptions) {
+						const target: IHoverDelegateTarget = {
+							targetElements: [this],
+							dispose: () => { }
+						};
+						const resolvedTooltip = await tooltip;
+						if (resolvedTooltip) {
+							hoverOptions = { text: resolvedTooltip, target, anchorPosition: AnchorPosition.BELOW };
+						}
+					}
+					if (hoverOptions) {
+						if (mouseX !== undefined) {
+							(<IHoverDelegateTarget>hoverOptions.target).x = mouseX + 10;
+						}
+						hoverDelegate.showHover(hoverOptions);
```

Comment:
Do we need to do anything with the `IDisposable` this call returns?

Reply:
I don't think so. The hover service should be disposing it when the hover is hidden. From what I understand, the `IDisposable` is for if you want to hide it.